server/src/main/java/org/elasticsearch/script/field/WriteField.java

```diff
@@ -193,37 +237,105 @@ public WriteField set(Object value) {
         if (value instanceof NestedDocument doc) {
             throw new IllegalArgumentException("cannot set NestedDocument [" + doc.getDoc() + "] as path [" + path + "]");
         }
-        container.put(leaf, value);
+        setToContainer(value);
+        throwExceptionIfPresent();
         return this;
     }
 
+    @SuppressWarnings("unchecked")
+    private void setToContainer(Object value) {
+        clearErrors();
+        if (container instanceof Map<?, ?>) {
+            Map<String, Object> mapContainer = (Map<String, Object>) container;
+            mapContainer.put(leaf, value);
+        } else if (container instanceof List<?>) {
+            List<Object> listContainer = (List<Object>) container;
+            int index = parseIndex(leaf);
+            if (index >= 0 && index < listContainer.size()) {
+                listContainer.set(index, value);
+            } else {
```

Comment:
If this were handled in the spirit of the fields api, we would expand the list rather than error.  See [doc(int index)](https://github.com/elastic/elasticsearch/pull/96786/files#diff-f93361aeb2c3e4d052c5298f79e8cf885eb8d6750dd4e0d2723475707b793a7fR656)

Reply:
I've addressed this in 376edbb72a9d4a3b7556d09d5cc5d36615c10460